src/Workspaces/Core/Portable/Workspace/Solution/SolutionState.cs

```diff
@@ -1549,12 +1549,27 @@ public ImmutableArray<DocumentId> GetDocumentIdsWithFilePath(string? filePath)
 
         private ImmutableDictionary<ProjectId, ICompilationTracker> CreateCompilationTrackerMap(ProjectId changedProjectId, ProjectDependencyGraph dependencyGraph)
         {
-            var builder = ImmutableDictionary.CreateBuilder<ProjectId, ICompilationTracker>();
+            if (_projectIdToTrackerMap.Count == 0)
+                return _projectIdToTrackerMap;
 
+            var newTrackerInfo = ArrayBuilder<KeyValuePair<ProjectId, ICompilationTracker>>.GetInstance();
+            var allReused = true;
```

Comment:
Are we better off just doing a loop first checking the dependency graph CanReuse() for each of the entries, and if so just bailing entirely? Otherwise here we're still allocating the builder only to throw it away.

Reply:
From what I've seen, pretty much everything about ImmutableDictionarys are slow, so I figured it was just better to make a single pass through it. The intent here is that the ArrayBuilder won't actually allocate normally, so the only real allocation that will happen is if the ImmutableDictionary really needs to be created.